### PR TITLE
perf(txindex): search optimization

### PR DESCRIPTION
**Pull request by inon-man**
This PR addresses several optimization points for TxIndex.Search:

1. Remove expensive map access from sorting function
2. Replace sort.Slice to sort.Sort with sorter approach
3. Reduce heap allocations from extractEventSeqFromKey

benchstat result:
```
goos: darwin
goarch: amd64
pkg: github.com/cometbft/cometbft/state/txindex/kv
cpu: Intel(R) Core(TM) i9-10910 CPU @ 3.60GHz
                     │   old.txt    │              new.txt               │
                     │    sec/op    │   sec/op     vs base               │
TxSearchBigResult-20   3236.1m ± 7%   945.9m ± 9%  -70.77% (p=0.002 n=6)

                     │   old.txt    │               new.txt               │
                     │     B/op     │     B/op      vs base               │
TxSearchBigResult-20   426.1Mi ± 0%   311.6Mi ± 0%  -26.86% (p=0.002 n=6)

                     │   old.txt   │              new.txt               │
                     │  allocs/op  │  allocs/op   vs base               │
TxSearchBigResult-20   4.040M ± 0%   2.040M ± 0%  -49.51% (p=0.002 n=6)
```

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: state/txindex/kv/kv.go [hunk 1]
```diff
@@ -27,9 +27,10 @@ import (
 )
 
 const (
-	tagKeySeparator     = "/"
-	tagKeySeparatorRune = '/'
-	eventSeqSeparator   = "$es$"
+	tagKeySeparator          = "/"
```
Comment (melekes) [inline state/txindex/kv/kv.go:30]: Do we still need this?

**Review by melekes — APPROVED** [review]
Thanks @inon-man ❤️

**Comment by ValarDragon** [comment]
Really exciting, thank you! Can we backport this?

File: state/txindex/kv/kv.go (hunk 1 above)
Comment (inon-man) [reply state/txindex/kv/kv.go:30]: I guess it is for conditional search.

**Comment by melekes** [comment]
> Really helpful, thank you! Can we backport this?

According to our backporting policy, we will only backport bug fixes to v0.38 and v0.37. We can backport to the `v0.37.x-experimental` and/or `v0.38.x-experimental` branches if needed.

**Pull request merged**